XMonad/Actions/DynamicWorkspaceNames.hs

```diff
@@ -0,0 +1,73 @@
+{-# LANGUAGE TypeSynonymInstances, MultiParamTypeClasses, FlexibleInstances #-}
+-----------------------------------------------------------------------------
+-- |
+-- Module      :  XMonad.Actions.DynamicWorkspaceNames
+-- Copyright   :  (c) Ivan A. Malison
+-- License     :  BSD-style (see LICENSE)
+--
+-- Maintainer  :  <IvanMalison@gmail.com>
+-- Stability   :  unstable
+-- Portability :  unportable
+--
+-- Automatically label workspaces using the WorkspaceNamesModule.
+-----------------------------------------------------------------------------
+module XMonad.Actions.DynamicWorkspaceNames
+  (
+   NameWorkspacesHook(..),
+   currentWSName,
+   updateWorkspaceName,
+   setWorkspaceNames,
+   getClass,
+   getNamesFromWindows,
+   getClassesNameForWorkspace
+  ) where
+
+import           Data.List
+import           Data.Maybe
+import           Graphics.X11.Xlib.Extras
+import           Control.Monad
+
+import           XMonad
+import qualified XMonad.StackSet as W
+import           XMonad.Actions.WorkspaceNames
+import           XMonad.Layout.LayoutModifier
+
+-- | Get the current name for the given workspace.
+currentWSName :: WindowSpace -> X String
+currentWSName ws = fromMaybe "" <$> (getWorkspaceNames' <*> pure (W.tag ws))
+
+-- | Update the name of a workspace only if it is not the same as the current
+-- name.
+updateWorkspaceName :: (WindowSpace -> X String) -> WindowSpace -> X ()
+updateWorkspaceName nameWorkspace workspace = do
+  currentName <- currentWSName workspace
+  newName <- nameWorkspace workspace
+  when (currentName /= newName) $ setWorkspaceName (W.tag workspace) newName
```

Comment:
This "only set if changed" functionality could go into `setWorkspaceName` itself.
And I feel like these `currentWSName`, `updateWorkspaceName` and `setWorkspaceNames` functions belong into `X.A.WorkspaceNames` rather than here.